overcooked_ai_py/agents/benchmarking.py

```diff
@@ -20,46 +20,71 @@ class AgentEvaluator(object):
     pickleable. We should think about possible improvements/what makes most sense to do here.
     """
 
-    def __init__(self, mdp_params, env_params={}, mdp_fn_params=None, force_compute=False, mlp_params=NO_COUNTERS_PARAMS, debug=False):
+    def __init__(self, mdp_params_lst, env_params={}, mdp_fn_params_lst=None, force_compute=False, mlp_params_lst=None, debug=False):
         """
-        mdp_params (dict): params for creation of an OvercookedGridworld instance through the `from_layout_name` method
+        mdp_params_lst (list): a list of params for creation of an OvercookedGridworld instance
+            through the `from_layout_name` method
         env_params (dict): params for creation of an OvercookedEnv
-        mdp_fn_params (dict): params to setup random MDP generation
+        mdp_fn_params (list): a list of params to setup random MDP generation
         force_compute (bool): whether should re-compute MediumLevelPlanner although matching file is found
         mlp_params (dict): params for MediumLevelPlanner
         """
-        assert type(mdp_params) is dict, "mdp_params must be a dictionary"
-
-        if mdp_fn_params is None:
-            mdp = OvercookedGridworld.from_layout_name(**mdp_params)
-            self.mdp_fn = lambda: mdp
-            self.env = OvercookedEnv.from_mdp(mdp, **env_params)
+        assert type(mdp_params_lst) is list and len(mdp_params_lst) > 0, \
+            "please make sure mdp_params_lst is a list. " \
+            "If not, please see where the function is called and make a fake list"
+        assert all([type(mdp_params_i) is dict for mdp_params_i in mdp_params_lst]), \
+            "all mdp_params must be a dictionary"
+
+        # The main changes here is that we are putting a list of environment in the AgentEvaluator.
```

Comment:
Just to clarify, what I'm suggesting is to always have 1 Environment object that is capable of dealing with all the complexity of multiple layouts within its code (i.e. same environment object has different mdps at each reset, rather than having a bunch of fixed-layout Envs).

Reply:
If it's possible to do this elegantly, it will pay off even for the rllib training itself (for which, again, we will only need 1 environment rather than keeping track of many).

Reply:
The main problem with having 1 environment object is that I really cannot find a way to "step" in 1 environment but multiple layouts.

I read through the relevant files in rllib, and tried that for a pretty long time and had no results.

I think I will work on the LayoutGenerator first and come back to see if I can amend this

Reply:
Hmm I see – in the overcooked_ai portion of the code that should be relatively straightforward and it worked fine for baselines, but yeah, I'm not sure if it would cause complications for rllib (although i would be surprised if one couldn't get around it).

Reply:
rllib sealed off a lot of functionality: I actually cannot even edit the train function in rllib

After poking around for a long time, I realize the only API we have control over is via the step function in the OvercookedEnv. Techinically that is still only 1 OvercookedEnv with many layout_env inside.